src/java/org/apache/cassandra/io/sstable/Descriptor.java

```diff
@@ -423,6 +425,19 @@ public boolean isCompatible()
         return version.isCompatible();
     }
 
+    public Set<Component> discoverComponents()
+    {
+        Set<Component.Type> knownTypes = Sets.difference(Component.TYPES, Collections.singleton(Component.Type.CUSTOM));
+        Set<Component> components = Sets.newHashSetWithExpectedSize(knownTypes.size());
+        for (Component.Type componentType : knownTypes)
+        {
+            Component component = new Component(componentType);
```

Comment:
Nit: Move the declaration of `singletons` after the components -- this avoids making people wonder how it's populated.

Reply:
I don't think I can do that because in that case when `new Component` was called, the `singletons` field would not be initialized yet.

Given your other comments, we will probably go back to `Component` class when making sstable format really pluggable